pycbc/inference/models/hierarchical.py

```diff
@@ -599,3 +605,227 @@ def _loglikelihood(self):
+                               marginalize_phase=marginalize_phase,
+                               **kwargs)
+
+    def write_metadata(self, fp, group=None):
+        """Adds metadata to the output files
+
+        Parameters
+        ----------
+        fp : pycbc.inference.io.BaseInferenceFile instance
+            The inference file to write to.
+        group : str, optional
+            If provided, the metadata will be written to the attrs specified
+            by group, i.e., to ``fp[group].attrs``. Otherwise, metadata is
+            written to the top-level attrs (``fp.attrs``).
+        """
+        super().write_metadata(fp, group=group)
+        sampattrs = fp.getattrs(group=fp.samples_group)
+        # if a group is specified, prepend the lognl names with it
+        if group is None or group == '/':
+            prefix = ''
+        else:
+            prefix = group.replace('/', '__')
+            if not prefix.endswith('__'):
+                prefix += '__'
+        try:
+            model = self.submodels[self.primary_model]
+            sampattrs['{}lognl'.format(prefix)] = model.lognl
+        except AttributeError:
+            pass
+
+    def _loglr(self):
+        r"""Computes the log likelihood ratio,
+
+        .. math::
+
+            \log \mathcal{L}(\Theta) = \sum_i
+                \left<h_i(\Theta)|d_i\right> -
+                \frac{1}{2}\left<h_i(\Theta)|h_i(\Theta)\right>,
+
+        at the current parameter values :math:`\Theta`.
+
+        Returns
+        -------
+        float
+            The value of the log likelihood ratio.
+        """
+        # calculate <d-h|d-h> = <h|h> - 2<h|d> + <d|d> up to a constant
+
+        # note that for SOBHB signals, ground-based detectors dominant SNR
+        # and accuracy of (tc, ra, dec)
+        sh_ground, hh_ground = self.primary_model.loglr(just_sh_hh=True)
+
+        # add likelihood contribution from space-borne detectors, we
+        # calculate sh/hh for each extrinsic parameter
+
+        logging.info("Calculating sh/hh for space-borne detectors")
+        parameter_set = self.primary_model.current_params
+        nums = self.primary_model.marginalize_vector_samples
+        sh_space = numpy.zeros(nums)
+        hh_space = numpy.zeros(nums)
+        for parameters in parameter_set:
+            i = numpy.where(parameter_set==parameters)[0][0]
+            self.other_models.update(**parameters)
+            sh_space[i], hh_space[i] = self.other_models.loglr(
+                                            just_sh_hh=True)
+
+        sh_total = sh_ground + sh_space
+        hh_total = hh_ground + hh_space
+        loglr = self.primary_model.marginalize_loglr(sh_total, hh_total)
+        return loglr
+
+    def _loglikelihood(self):
+        other_models_lognl = 0
+        for lbl, model in self.submodels.items():
+            # Update the parameters of each
+            model.update(**{p.subname: self.current_params[p.fullname]
```

Comment:
I think this may not work correctly. You should only update the variable parameters of the primary model.

Reply:
There are no errors from here in my test runs, but I will ensure that.